### ISSUE TITLE
Mongoose Discriminator key - Array column

### ISSUE THREAD

**Issue by redgumnfp** [Automattic/mongoose#10303]
**Do you want to request a *feature* or report a *bug*?**
Bug, previously mentioned:
https://github.com/Automattic/mongoose/issues/10130 (2nd message)

**What is the current behavior?**
When utilising an array column as the discriminator key, creating (and updating) documents with the key results in a double-nested array.

For example:
This:
`['123456', 'ABC123']`
Becomes this:
`[['123456'], ['ABC123']]`

**If the current behavior is a bug, please provide the steps to reproduce.**
Schema:
```
const mongooseSchema = new mongoose.Schema({
  tenantRefs: [{
    type: String,
    unique: false,
    required: true,
    index: true,
  }],
  otherData: {
    type: String,
    unique: false,
    required: true,
    index: true,
  },
  ...  
});
```

This can then be consumed with a dynamic model, created at run-time:
```
const tenantWrapper = (name, schema, options) => () => {
  // pull these from CLS, or pass in manually
  const tenantValue = getTenantValue();
  const skipTenant = getSkipTenant() || false;

  const Model = connectionRegional.model(name, schema);
  if (skipTenant) {
    Model.schema.set('discriminatorKey', null);
    return Model;
  }


  const discriminatorName = `${name}-${tenantValue}`;
  schema.set('discriminatorKey', 'tenantRefs');

  const existingDiscriminator = (Model.discriminators || {})[discriminatorName];
  if (existingDiscriminator) return existingDiscriminator;


  return Model.discriminator(
    discriminatorName,
    new mongoose.Schema({}, options),
    tenantValue,
  );
};
```

In essence, I'm attempting to utilise the tenantRefs array to automatically filter depending on the user's tenantRef. If you can imagine a marketplace scenario, there's an array including a supplier ID and a client ID.

This works perfectly on all query/aggregation methods, but frustratingly, not when creating a document.

Due to this snippet in /lib/document.js, the array is split out:

```
if (this.$__schema.singleNestedPaths[path] == null) {
      // If this path is underneath a single nested schema, we'll call the setter
      // later in `$__set()` because we don't take `_doc` when we iterate through
      // a single nested doc. That's to make sure we get the correct context.
      // Otherwise we would double-call the setter, see gh-7196.
      val = schema.applySetters(val, this, false, priorVal);
    }
```

**What is the expected behavior?**
When creating documents with an array discriminator key, either of the below should happen:

1. A validation error could be shown that the value cannot be set due to the discriminator key being an array type (helps reduce dev debugging, but doesn't fix the problem)
2. The discriminator key field should be set correctly to the value provided (i.e. no transformation)

**What are the versions of Node.js, Mongoose and MongoDB you are using? Note that "latest" is not a version.**
Node v14.16.0
Mongoose v5.12.11
MongoDb v4.4.4

**Comment by vkarpov15** [comment]
How are you creating the document? I just get an empty array for `tenantRefs` with the below code:

```javascript
'use strict';

const mongoose = require('mongoose');

mongoose.set('debug', true);
mongoose.set('useFindAndModify', false);

const { Schema } = mongoose;

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });

  await mongoose.connection.dropDatabase();

  const mongooseSchema = new mongoose.Schema({
    tenantRefs: [{
      type: String,
      required: true,
    }],
    otherData: {
      type: String,
      required: true,
    },
  }, { discriminatorKey: 'tenantRefs' });
  const Model = mongoose.model('Test', mongooseSchema);

  const D = Model.discriminator(
    'D',
    new mongoose.Schema({}),
    ['abc', '123'],
  );

  const res = await D.create({ otherData: 'test' });
  console.log(res); // tenantRefs: []
}

```

**Comment by redgumnfp** [comment]
@vkarpov15 - thanks for picking this up 💪

When creating a model, I would pass in the discriminator value.

So on your example script, the below:
```
const res = await D.create({ 
    tenantRefs: ['abc', '123'],
    otherData: 'test',
});
```

Would actually create the below in error:
```
{
    tenantRefs: [['abc'], ['123']],
    otherData: 'test',
}
```

Hope that helps, but please let me know if you can replicate or not.

**Comment by vkarpov15** [comment]
@redgumnfp unfortunately I haven't been able to repro. I get an empty array for `tenantRefs`. I'll try again and see if the Mongoose version makes a difference, I was probably using 5.12.13.

**Comment by vkarpov15** [comment]
Below script works with 35d2041. Fix will be in v5.13.1 :+1:

```javascript
'use strict';
  
const mongoose = require('mongoose');

mongoose.set('debug', true);
mongoose.set('useFindAndModify', false);

const { Schema } = mongoose;

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test', {
    useNewUrlParser: true,
    useUnifiedTopology: true
  });

  await mongoose.connection.dropDatabase();

  const mongooseSchema = new mongoose.Schema({
    tenantRefs: [{
      type: String,
      required: true,
    }],
    otherData: {
      type: String,
      required: true,
    },
  }, { discriminatorKey: 'tenantRefs' });
  const Model = mongoose.model('Test', mongooseSchema);

  const D = Model.discriminator(
    'D',
    new mongoose.Schema({}),
    ['abc', '123'],
  );

  const res = await D.create({ otherData: 'test', tenantRefs: ['abc', '123'] });
  console.log(res);

  console.log(await D.findOne());
}

```

**Comment by redgumnfp** [comment]
Great, thank you @vkarpov15 👏👏👏

I've make a note to review once v5.13.1 is released.
